Spigot-API-Patches/0186-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,84 @@
+From: slicklibro <slicklibro@gmail.com>
+Date: Sat, 28 Sep 2019 00:01:23 +1200
+Subject: [PATCH] Ticket-level per player view distance API
+
+
+diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
+index 8d8d8957..ae2fe927 100644
+--- a/src/main/java/org/bukkit/entity/Player.java
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -1860,18 +1860,65 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
+     public void setAffectsSpawning(boolean affects);
+ 
+     /**
+-     * Gets the view distance for this player
++     * Gets the ticket view distance for this player
++     * <br>
++     * Returns the ticket view distance.
+      *
+-     * @return the player's view distance
++     * @return the player's ticket view distance
+      */
++    @Deprecated
+     public int getViewDistance();
+ 
+     /**
+-     * Sets the view distance for this player
+-     *
+-     * @param viewDistance the player's view distance
++     * Sets both the ticket and tracking view distance for this player
++     * <br>
++     * This will update both the ticket and tracking view distance value.
++     * Setting both at the same value will disallow shared view distance.
++     * 
++     * @param viewDistance the value to set both the player's ticket & tracking view distance
+      */
+     public void setViewDistance(int viewDistance);
++
++    /**
++     * Gets the ticket view distance for this player
++     * <br>
++     * Returns the ticket view distance.
++     *
++     * @return the player's ticket view distance
++     */
++    public int getTicketViewDistance();
++
++    /**
++     * Sets only the ticket view distance for this player
++     * <br>
++     * This determines how far the player can add tickets to chunks.
++     * This is bound under the tracking view distance.
++     *
++     * @param ticketViewDistance the player's ticket view distance
++     */
++    public void setTicketViewDistance(int ticketViewDistance);
++
++    /**
++     * Gets the tracking view distance for this player
++     * <br>
++     * This determines how far the player can see/load all chunks with tickets.
++     * If this is higher than the ticket view distance, the player can track chunks with tickets added by other players.
++     * Allowing shared view distance.
++     *
++     * @return the player's tracking view distance
++     */
++    public int getTrackViewDistance();
++
++    /**
++     * Sets the tracking view distance for this player
++     * <br>
++     * This determines how far the player can see/load all chunks with tickets.
++     * If this is higher than the ticket view distance, the player can track chunks with tickets added by other players.
++     * Allowing shared view distance.
++     *
++     * @param trackViewDistance the player's tracking view distance
++     */
++    public void setTrackViewDistance(int trackViewDistance);
```

Comment:
The problem with this API is that it's not easy to maintain and implement correctly. The api also has a narrow use case, which is to share chunks where a lot of players are (i.e spawn). Currently it's very possible for plugins to simply increase view distances in such areas. On that basis alone I can't support this additional API.

However for the implementation there are concerns which are NOT met:

1. Ticket view distance must be used for the entity tracker. You CANNOT go over that value UNLESS you track what players have what chunks loaded. We should not send entities into unloaded chunks.
2. You _must_ propagate block changes outside the ticket range. Currently your code does NOT allow this to occur.
Chunks only update to players when they're at entity-ticking status. With your changes, "tracking" chunks are not going to be at this status due to a lack of player tickets.
3. Your current documentation is poor and does not remotely describe how a tracking view distance differs from a ticket view distance.

As a side note, your current implementation does NOT prevent chunks outside the tracking range from being FULLY re-sent to clients once they get into the ticket range. So there's a waste of resources right there, however it would be required to do this given your implementation's lack of correctly sending block updates to clients.

Another point is just how janky this looks to clients. The chunks don't load evenly (since the unload distance is not equal to the load distance), and there certainly will be gaps when you consider the situation where a client has a large enough tracking range to peek at another player's chunks, but since the tracking range > ticket range, there would be an obvious gap inbetween where chunks simply aren't loaded.

Reply:
I'll find solutions to your first two issues, and I'll update the documentation to be more precise and easier to understand. There is no possible way I am closing the door on the option of shared view distance. Your last points on 'how janky this looks to clients' is simply an opinion, I know a slew of server owners that have been wanting this for a long time. Many people would greatly prefer extension of their view distance instead of just constantly sitting on 4-6 all the time, which is a terrible gameplay experience. Many players also do not play alone, they play in groups with friends, and these people will benefit massively (think about factions, etc.). This is definitely not a specific use case for places just like spawn...

This is a major benefit, and I am not dropping it, especially with this amount of structure built in already. I'll get back to you when these solutions are committed.

Reply:
Unless it is out of spite, I have no idea why you would be against shared view distance. It is left as an option and not as something that is compulsory. The point about it looking janky to clients is obtuse, you literally on the client side cache chunks that you load behind, if I walked in a full circle or zigzag on a normal server I would produce the exact same effect you are describing with literally no difference.

Once the documentation is updated to be more precise and understandable, we can test this extensively. It would be a major blunder and waste to not have this option.

Reply:
You have misunderstood what the chunk tracking distance means, the name is from vanilla mappings as instead of `#sendChunk` the method is named `#updateChunkTracking`, so I named it as such (as it makes much more sense). All the chunk tracking distance does is determine the update distance for chunk tracking, sending chunks within range in `#getVisibleChunks` (which is determined by `updatingChunks` which is determined by `ChunkMapDistance`, which we have modified to for per player view distance tickets). In other words, the point of it is to determine the range where we send chunks with player tickets attached. It simply sets the encompassing distance for displaying chunks with player tickets. The logic is equivalent to mojang's `spectatorsGenerateChunks`

Reply:
After another analysis of the entity tracking view distance management, your previous note to not use the chunk tracking view distance is incorrect. 

Using it the tracking view distance is the completely correct method, as it performs a check against `#getVisibleChunk`, which again we control through our already implemented logic at the base level. It is extremely important to remember that we control the list of chunks with player tickets. Much of this logic has already been handled at the base level.

Reply:
1. 
Chunks outside the ticket range are kept loaded to clients if tracking range > ticket range. If you inspect the code that actually broadcasts block updates to clients, you'll see that it's only invoked for chunks at entity ticking level. Your code does not ensure that block updates actually get propagated to clients in this case.


2.
Re-read my post if you think spite is the only reason I could be against this.

3. 
See 1.
Note: Chunks can be sent to players if they're at ticking level, they're not required to have player tickets.

4.
I'm also aware of this, and it doesn't change my original point. That code will only execute if the chunk is at ticking level, and I care about the case where it's not.

Reply:
If you are talking about the gap in between, we can modify the code to untrack correctly against #getVisibleChunks

Reply:
what

Reply:
check the latest commit